src/System.Net.Utilities/src/System/Net/NetworkInformation/Ping.Unix.cs

```diff
@@ -2,33 +2,275 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System.ComponentModel;
+using System.Diagnostics;
+using System.IO;
+using System.Net.Sockets;
+using System.Runtime.InteropServices;
+using System.Text;
 using System.Threading;
+using System.Threading.Tasks;
 
 namespace System.Net.NetworkInformation
 {
     public partial class Ping
     {
-        private void InternalSendAsync(IPAddress address, byte[] buffer, int timeout, PingOptions options)
+        private const int IcmpHeaderLengthInBytes = 8;
+        private const int IpHeaderLengthInBytes = 20;
+
+        private static readonly string s_ipv4PingFile = "ping";
+        private static readonly string s_ipv6PingFile = "ping6";
+
+        private async void InternalSendAsync(IPAddress address, byte[] buffer, int timeout, PingOptions options)
         {
             AsyncOperation asyncOp = _asyncOp;
             SendOrPostCallback callback = _onPingCompletedDelegate;
 
-            // TODO: Implement this (#2487)
+            PingReply pr = null;
+            Exception pingException = null;
+
+            try
+            {
+                if (RawSocketPermissions.CanUseRawSockets())
+                {
+                    pr = await SendIcmpEchoRequestOverRawSocket(address, buffer, timeout, options);
+                }
+                else
+                {
+                    pr = await SendWithPingUtility(address, buffer, timeout, options);
```

Comment:
If you wanted, you could also make it a single await/ConfigureAwait(false), e.g.
```C#
pr = await (RawSocketPermissions.CanUseRawSockets() ?
    SendIcmpEchoRequestOverRawSocket(address, buffer, timeout, options) :
    SendWithPingUtility(address, buffer, timeout, options)).ConfigureAwait(false);
```